pinot-core/src/main/java/org/apache/pinot/core/plan/PinotExplainedRelNode.java

```diff
@@ -0,0 +1,146 @@
+ * (aka {@link org.apache.pinot.core.operator.BaseOperator}), which include important information like whether indexes
+ * are being used or not.
+ *
+ * {@link PinotExplainedRelNode} is a way to represent these single-stage operators in Calcite.
+ * <b>They are not meant to be executed</b>.
+ * Instead, when physical explain is required, the broker ask for the final plan to each server. They return
+ * a PlanNode and the broker then converts these PlanNodes into a PinotExplainedRelNode. The logical plan is then
+ * modified to substitute the nodes that were converted into single-stage physical plans with the corresponding
+ * PinotExplainedRelNode.
+ */
+public class PinotExplainedRelNode extends AbstractRelNode {
+
+  /**
+   * The name of this node, whose role is like a title in the explain plan.
+   */
+  private final String _type;
+  private final Map<String, Plan.ExplainNode.AttributeValue> _attributes;
+  private final List<RelNode> _inputs;
+  private final DataSchema _dataSchema;
+
+  public PinotExplainedRelNode(RelOptCluster cluster, RelTraitSet traitSet, String type,
+      Map<String, Plan.ExplainNode.AttributeValue> attributes, DataSchema dataSchema, RelNode input) {
+    this(cluster, traitSet, type, attributes, dataSchema, Lists.newArrayList(input));
+  }
+
+  public PinotExplainedRelNode(RelOptCluster cluster, RelTraitSet traitSet, String type,
+      Map<String, Plan.ExplainNode.AttributeValue> attributes, DataSchema dataSchema, List<? extends RelNode> inputs) {
+    super(cluster, traitSet);
+    _type = type;
+    _attributes = attributes;
+    _inputs = new ArrayList<>(inputs);
+    _dataSchema = dataSchema;
+  }
+
+  @Override
+  protected RelDataType deriveRowType() {
+    int size = _dataSchema.size();
+    List<RelDataTypeField> fields = new ArrayList<>(size);
+    RelDataTypeFactory typeFactory = getCluster().getTypeFactory();
+    for (int i = 0; i < size; i++) {
+      String columnName = _dataSchema.getColumnName(i);
+
+      DataSchema.ColumnDataType columnDataType = _dataSchema.getColumnDataType(i);
+      RelDataType type = columnDataType.toType(typeFactory);
+
+      fields.add(new RelDataTypeFieldImpl(columnName, i, type));
+    }
+    return new RelRecordType(StructKind.FULLY_QUALIFIED, fields, false);
+  }
+
+  @Override
+  public List<RelNode> getInputs() {
+    return Collections.unmodifiableList(_inputs);
+  }
+
+  @Override
+  public void replaceInput(int ordinalInParent, RelNode p) {
+    _inputs.set(ordinalInParent, p);
+  }
+
+  @Override
+  public String getRelTypeName() {
+    return _type;
+  }
+
+  @Override
+  public RelWriter explainTerms(RelWriter pw) {
+    RelWriter relWriter = super.explainTerms(pw);
+    int inputSize = getInputs().size();
+    if (inputSize == 1) {
+      relWriter.input("input", getInputs().get(0));
+    } else if (inputSize > 1) {
+      for (int i = 0; i < inputSize; i++) {
+        relWriter.input("input#" + i, getInputs().get(i));
+      }
+    }
```

Comment:
> Where is this input term string being used?

It is used internally in RelWriter (in Calcite).

> Does it matter if for a single input node we use input#0 instead of input?

Not really. As far as I can tell the only requirement is that inputs should have a unique name. So each input should be have a different `input#<id>`. How this name is used depends on the Calcite printer and as far as I can see neither of them actually print this name